File: Go/z_algorithm_utils/z_algorithm_utils_test.go
// Unit tests for Z Algorithm Utilities (Go version)

package zalgorithmutils

import (
	"testing"
)

// ============================================================================
// Z-Array Tests
// ============================================================================

func TestZArrayBasic(t *testing.T) {
	// "aabcaabxaaz" -> [0, 1, 0, 0, 3, 1, 0, 0, 2, 1, 0]
	result := ZArray("aabcaabxaaz")
	expected := []int{0, 1, 0, 0, 3, 1, 0, 0, 2, 1, 0}
	if len(result) != len(expected) {
		t.Errorf("Length mismatch: got %d, want %d", len(result), len(expected))
	}
	for i := range expected {
		if result[i] != expected[i] {
			t.Errorf("Z[%d]: got %d, want %d", i, result[i], expected[i])
		}
	}
}

func TestZArrayEmpty(t *testing.T) {
	result := ZArray("")
	if len(result) != 0 {
		t.Errorf("Empty string should return empty array, got %d elements", len(result))
	}
}

func TestZArraySingleChar(t *testing.T) {
	result := ZArray("a")
	if len(result) != 1 || result[0] != 0 {
		t.Errorf("Single char should return [0], got %v", result)
	}
}

func TestZArrayAllSame(t *testing.T) {
	// "aaaa" -> [0, 3, 2, 1]
	result := ZArray("aaaa")
	expected := []int{0, 3, 2, 1}
	for i := range expected {
		if result[i] != expected[i] {
			t.Errorf("Z[%d]: got %d, want %d", i, result[i], expected[i])
		}
	}
}

func TestZArrayNoMatches(t *testing.T) {
	result := ZArray("abcd")
	for i := 1; i < len(result); i++ {
		if result[i] != 0 {
			t.Errorf("Z[%d]: got %d, want 0", i, result[i])
		}
	}
}

func TestZArrayBytes(t *testing.T) {
	data := []byte("aabcaabxaaz")
	result := ZArrayBytes(data)
	expected := []int{0, 1, 0, 0, 3, 1, 0, 0, 2, 1, 0}
	for i := range expected {
		if result[i] != expected[i] {
			t.Errorf("Z[%d]: got %d, want %d", i, result[i], expected[i])
		}
	}
}

// ============================================================================
// Pattern Matching Tests
// ============================================================================

func TestFindAllOccurrences(t *testing.T) {
	result := FindAllOccurrences("abc", "abcabcabc")
	expected := []int{0, 3, 6}
	if len(result) != len(expected) {
		t.Errorf("Length mismatch: got %d, want %d", len(result), len(expected))
	}
	for i := range expected {
		if result[i] != expected[i] {
			t.Errorf("Position[%d]: got %d, want %d", i, result[i], expected[i])
		}
	}
}

func TestFindAllOccurrencesNotFound(t *testing.T) {
	result := FindAllOccurrences("xyz", "abcabcabc")
	if len(result) != 0 {
		t.Errorf("Should find no occurrences, got %d", len(result))
	}
}

func TestFindFirstOccurrence(t *testing.T) {
	tests := []struct {
		pattern string
		text    string
		want    int
	}{
		{"abc", "xyzabc", 3},
		{"abc", "abcabc", 0},
		{"abc", "xyz", -1},
		{"", "abc", -1},
		{"abc", "", -1},
	}

	for _, tt := range tests {
		got := FindFirstOccurrence(tt.pattern, tt.text)
		if got != tt.want {
			t.Errorf("FindFirstOccurrence(%s, %s): got %d, want %d", tt.pattern, tt.text, got, tt.want)
		}
	}
}

func TestCountOccurrences(t *testing.T) {
	tests := []struct {
		pattern string
		text    string
		want    int
	}{
		{"a", "banana", 3},
		{"ana", "banana", 2},
		{"xyz", "abcabc", 0},
	}

	for _, tt := range tests {
		got := CountOccurrences(tt.pattern, tt.text)
		if got != tt.want {
			t.Errorf("CountOccurrences(%s, %s): got %d, want %d", tt.pattern, tt.text, got, tt.want)
		}
	}
}

func TestFindMatches(t *testing.T) {
	matches := FindMatches("abc", "abcabc")
	if len(matches) != 2 {
		t.Errorf("Should find 2 matches, got %d", len(matches))
	}
	if matches[0].Index != 0 || matches[0].Length != 3 {
		t.Errorf("First match: got index=%d, length=%d, want 0, 3", matches[0].Index, matches[0].Length)
	}
	if matches[1].Index != 3 || matches[1].Length != 3 {
		t.Errorf("Second match: got index=%d, length=%d, want 3, 3", matches[1].Index, matches[1].Length)
	}
}

// ============================================================================
// Substring Analysis Tests
// ============================================================================

func TestLongestPrefixSuffix(t *testing.T) {
	tests := []struct {
		s    string
		want int
	}{
		{"ababa", 3},   // "aba"
		{"aaaa", 3},    // "aaa"
		{"abcd", 0},
		{"", 0},
	}

	for _, tt := range tests {
		got := LongestPrefixSuffix(tt.s)
		if got != tt.want {
			t.Errorf("LongestPrefixSuffix(%s): got %d, want %d", tt.s, got, tt.want)
		}
	}
}

func TestLongestRepeatedSubstring(t *testing.T) {
	substr, positions := LongestRepeatedSubstring("banana")
	if substr != "ana" {
		t.Errorf("Got substr=%s, want 'ana'", substr)
	}
	found := false
	for _, pos := range positions {
		if pos == 1 || pos == 3 {
			found = true
		}
	}
	if !found {
		t.Errorf("Positions should include 1 or 3, got %v", positions)
	}
}

func TestLongestCommonPrefix(t *testing.T) {
	tests := []struct {
		s1   string
		s2   string
		want int
	}{
		{"abcdef", "abcxyz", 3},
		{"xyz", "abc", 0},
		{"same", "same", 4},
	}

	for _, tt := range tests {
		got := LongestCommonPrefix(tt.s1, tt.s2)
		if got != tt.want {
			t.Errorf("LongestCommonPrefix(%s, %s): got %d, want %d", tt.s1, tt.s2, got, tt.want)
		}
	}
}

// ============================================================================
// Period Detection Tests
// ============================================================================

func TestFindMinimalPeriod(t *testing.T) {
	period := FindMinimalPeriod("abcabcabc")
	if period.Period != 3 {
		t.Errorf("Got period=%d, want 3", period.Period)
	}
	if !period.IsPeriodic {
		t.Errorf("Should be periodic")
	}

	period = FindMinimalPeriod("abcde")
	if period.Period != 5 {
		t.Errorf("Got period=%d, want 5", period.Period)
	}
	if period.IsPeriodic {
		t.Errorf("Should not be periodic")
	}
}

func TestIsRotation(t *testing.T) {
	tests := []struct {
		s1   string
		s2   string
		want bool
	}{
		{"abcde", "cdeab", true},
		{"abc", "abc", true},
		{"abcde", "abced", false},
		{"abc", "abcd", false},
	}

	for _, tt := range tests {
		got := IsRotation(tt.s1, tt.s2)
		if got != tt.want {
			t.Errorf("IsRotation(%s, %s): got %v, want %v", tt.s1, tt.s2, got, tt.want)
		}
	}
}

func TestFindAllRotations(t *testing.T) {
	rotations := FindAllRotations("abc")
	if len(rotations) != 3 {
		t.Errorf("Should have 3 rotations, got %d", len(rotations))
	}

	expected := []string{"abc", "bca", "cab"}
	for _, exp := range expected {
		found := false
		for _, rot := range rotations {
			if rot == exp {
				found = true
			}
		}
		if !found {
			t.Errorf("Missing rotation %s", exp)
		}
	}
}

// ============================================================================
// Compression Tests
// ============================================================================

func TestCompressString(t *testing.T) {
	pattern, count := CompressString("abcabcabc")
	if pattern != "abc" {
		t.Errorf("Got pattern=%s, want 'abc'", pattern)
	}
	if count != 3 {
		t.Errorf("Got count=%d, want 3", count)
	}

	pattern, count = CompressString("abcde")
	if pattern != "abcde" {
		t.Errorf("Got pattern=%s, want 'abcde'", pattern)
	}
	if count != 1 {
		t.Errorf("Got count=%d, want 1", count)
	}
}

func TestDecompressString(t *testing.T) {
	result := DecompressString("abc", 3)
	if result != "abcabcabc" {
		t.Errorf("Got %s, want 'abcabcabc'", result)
	}
}

// ============================================================================
// Similarity Tests
// ============================================================================

func TestSimilarityScore(t *testing.T) {
	tests := []struct {
		s1   string
		s2   string
		want float64
	}{
		{"abcdef", "abcxyz", 0.5},
		{"same", "same", 1.0},
		{"abc", "xyz", 0.0},
		{"", "", 1.0},
		{"abc", "", 0.0},
	}

	for _, tt := range tests {
		got := SimilarityScore(tt.s1, tt.s2)
		if got != tt.want {
			t.Errorf("SimilarityScore(%s, %s): got %f, want %f", tt.s1, tt.s2, got, tt.want)
		}
	}
}

func TestBatchSimilarity(t *testing.T) {
	base := "abcdef"
	strings := []string{"abcxyz", "abcdef", "xyz", ""}
	scores := BatchSimilarity(base, strings)
	if len(scores) != 4 {
		t.Errorf("Should have 4 scores, got %d", len(scores))
	}
}

// ============================================================================
// Utility Tests
// ============================================================================

func TestValidateZArray(t *testing.T) {
	s := "aabcaabxaaz"
	z := ZArray(s)
	if !ValidateZArray(s, z) {
		t.Errorf("Computed Z-array should be valid")
	}

	invalidZ := []int{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0}
	if ValidateZArray(s, invalidZ) {
		t.Errorf("Invalid Z-array should not be valid")
	}
}

func TestVisualizeZArray(t *testing.T) {
	vis := VisualizeZArray("aaaa")
	if vis == "" {
		t.Errorf("Visualization should not be empty")
	}
	if !contains(vis, "aaaa") {
		t.Errorf("Visualization should contain the string")
	}
}

func contains(s, substr string) bool {
	for i := 0; i <= len(s)-len(substr); i++ {
		if s[i:i+len(substr)] == substr {
			return true
		}
	}
	return false
}

// ============================================================================
// Pattern Matcher Tests
// ============================================================================

func TestZPatternMatcherSingle(t *testing.T) {
	matcher := NewZPatternMatcher([]string{"abc"})
	results := matcher.Search("abcabc")
	if len(results) != 2 {
		t.Errorf("Should find 2 matches, got %d", len(results))
	}
}

func TestZPatternMatcherMultiple(t *testing.T) {
	matcher := NewZPatternMatcher([]string{"ab", "bc"})
	results := matcher.Search("abcabc")
	if len(results) != 4 {
		t.Errorf("Should find 4 matches (2 'ab' + 2 'bc'), got %d", len(results))
	}
}

func TestZPatternMatcherSearchFirst(t *testing.T) {
	matcher := NewZPatternMatcher([]string{"bc", "ab"})
	result := matcher.SearchFirst("abcabc")
	if result == nil {
		t.Errorf("Should find first match")
	}
	if result.Position != 0 {
		t.Errorf("First match should be at position 0, got %d", result.Position)
	}
}

func TestZPatternMatcherCountAll(t *testing.T) {
	matcher := NewZPatternMatcher([]string{"a", "b", "c"})
	counts := matcher.CountAll("abcabc")
	if counts["a"] != 2 {
		t.Errorf("Count for 'a' should be 2, got %d", counts["a"])
	}
}

func TestZPatternMatcherNoMatch(t *testing.T) {
	matcher := NewZPatternMatcher([]string{"xyz"})
	results := matcher.Search("abcabc")
	if len(results) != 0 {
		t.Errorf("Should find no matches, got %d", len(results))
	}
	if matcher.SearchFirst("abcabc") != nil {
		t.Errorf("Should return nil for no match")
	}
}

// ============================================================================
// Helper Function Tests
// ============================================================================

func TestContains(t *testing.T) {
	if !Contains("abc", "abcabc") {
		t.Errorf("'abcabc' should contain 'abc'")
	}
	if Contains("xyz", "abcabc") {
		t.Errorf("'abcabc' should not contain 'xyz'")
	}
}

func TestReplaceAll(t *testing.T) {
	result := ReplaceAll("abc", "xyz", "abcabcabc")
	if result != "xyzxyzxyz" {
		t.Errorf("Got %s, want 'xyzxyzxyz'", result)
	}
}

func TestSplitByPattern(t *testing.T) {
	parts := SplitByPattern(",", "a,b,c")
	if len(parts) != 3 {
		t.Errorf("Should have 3 parts, got %d", len(parts))
	}
}

// ============================================================================
// Data Types Tests
// ============================================================================

func TestZMatchString(t *testing.T) {
	m := ZMatch{Index: 2, Length: 3, Text: "abcdef"}
	substr := m.String()
	if substr != "cde" {
		t.Errorf("Got %s, want 'cde'", substr)
	}
}

func TestStringPeriodPeriodString(t *testing.T) {
	p := StringPeriod{String: "abcabc", Period: 3, IsPeriodic: true}
	period := p.PeriodString()
	if period != "abc" {
		t.Errorf("Got %s, want 'abc'", period)
	}
}

File: Python/z_algorithm_utils/z_algorithm_utils_test.py
"""
Unit tests for Z Algorithm Utilities

Tests cover all major functions: Z-array computation, pattern matching,
substring analysis, period detection, and palindrome operations.
"""

import unittest
from typing import List
import sys
import os

# Add parent directory to path
sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from z_algorithm_utils.mod import (
    z_array, z_array_bytes, z_array_with_sentinel,
    find_all_occurrences, find_first_occurrence, count_occurrences, find_matches,
    longest_prefix_suffix, longest_repeated_substring, find_all_repeated_substrings,
    longest_common_prefix,
    find_minimal_period, is_rotation, find_all_rotations,
    longest_palindromic_prefix, longest_palindromic_suffix,
    similarity_score, batch_similarity,
    distinct_substring_count, compress_string, decompress_string,
    iter_occurrences,
    z_to_border, border_to_z,
    ZPatternMatcher, ZMatch, StringPeriod,
    visualize_z_array, validate_z_array
)


class TestZArray(unittest.TestCase):
    """Test Z-array computation."""
    
    def test_basic_z_array(self):
        """Test basic Z-array computation."""
        # "aabcaabxaaz" -> [0, 1, 0, 0, 3, 1, 0, 0, 2, 1, 0]
        result = z_array("aabcaabxaaz")
        expected = [0, 1, 0, 0, 3, 1, 0, 0, 2, 1, 0]
        self.assertEqual(result, expected)
    
    def test_empty_string(self):
        """Test Z-array for empty string."""
        self.assertEqual(z_array(""), [])
    
    def test_single_char(self):
        """Test Z-array for single character."""
        self.assertEqual(z_array("a"), [0])
    
    def test_all_same(self):
        """Test Z-array for string of same characters."""
        # "aaaa" -> [0, 3, 2, 1]
        self.assertEqual(z_array("aaaa"), [0, 3, 2, 1])
    
    def test_no_matches(self):
        """Test Z-array for string with no prefix matches."""
        # "abcd" -> [0, 0, 0, 0]
        self.assertEqual(z_array("abcd"), [0, 0, 0, 0])
    
    def test_pattern_in_pattern(self):
        """Test Z-array where pattern appears multiple times."""
        # "abcabcabc" -> [0, 0, 0, 6, 0, 0, 3, 0, 0]
        result = z_array("abcabcabc")
        expected = [0, 0, 0, 6, 0, 0, 3, 0, 0]
        self.assertEqual(result, expected)
    
    def test_z_array_bytes(self):
        """Test Z-array computation for bytes."""
        data = b"aabcaabxaaz"
        result = z_array_bytes(data)
        expected = [0, 1, 0, 0, 3, 1, 0, 0, 2, 1, 0]
        self.assertEqual(result, expected)
    
    def test_z_array_with_sentinel(self):
        """Test Z-array with sentinel."""
        pattern = "abc"
        text = "abcabc"
        result = z_array_with_sentinel(pattern, text)
        # Should find pattern occurrences
        self.assertTrue(any(z >= 3 for z in result[4:]))


class TestPatternMatching(unittest.TestCase):
    """Test pattern matching functions."""
    
    def test_find_all_occurrences(self):
        """Test finding all occurrences."""
        self.assertEqual(find_all_occurrences("abc", "abcabcabc"), [0, 3, 6])
        self.assertEqual(find_all_occurrences("ab", "ababab"), [0, 2, 4])
        self.assertEqual(find_all_occurrences("xyz", "abcabcabc"), [])
    
    def test_find_first_occurrence(self):
        """Test finding first occurrence."""
        self.assertEqual(find_first_occurrence("abc", "xyzabc"), 3)
        self.assertEqual(find_first_occurrence("abc", "abcabc"), 0)
        self.assertEqual(find_first_occurrence("abc", "xyz"), -1)
    
    def test_count_occurrences(self):
        """Test counting occurrences."""
        self.assertEqual(count_occurrences("a", "banana"), 3)
        self.assertEqual(count_occurrences("ana", "banana"), 2)
        self.assertEqual(count_occurrences("xyz", "abcabc"), 0)
    
    def test_empty_pattern_or_text(self):
        """Test with empty pattern or text."""
        self.assertEqual(find_all_occurrences("", "abc"), [])
        self.assertEqual(find_all_occurrences("abc", ""), [])
        self.assertEqual(find_first_occurrence("", "abc"), -1)
        self.assertEqual(find_first_occurrence("abc", ""), -1)
    
    def test_pattern_longer_than_text(self):
        """Test when pattern is longer than text."""
        self.assertEqual(find_all_occurrences("abcdef", "abc"), [])
    
    def test_find_matches(self):
        """Test finding matches with details."""
        matches = find_matches("abc", "abcabc")
        self.assertEqual(len(matches), 2)
        self.assertEqual(matches[0].index, 0)
        self.assertEqual(matches[0].length, 3)
        self.assertEqual(matches[0].matched_substring, "abc")
    
    def test_iter_occurrences(self):
        """Test occurrence iterator."""
        positions = list(iter_occurrences("abc", "abcabcabc"))
        self.assertEqual(positions, [0, 3, 6])


class TestSubstringAnalysis(unittest.TestCase):
    """Test substring analysis functions."""
    
    def test_longest_prefix_suffix(self):
        """Test longest prefix-suffix."""
        self.assertEqual(longest_prefix_suffix("ababa"), 3)  # "aba"
        self.assertEqual(longest_prefix_suffix("aaaa"), 3)  # "aaa"
        self.assertEqual(longest_prefix_suffix("abcd"), 0)
        self.assertEqual(longest_prefix_suffix(""), 0)
    
    def test_longest_repeated_substring(self):
        """Test longest repeated substring."""
        substr, positions = longest_repeated_substring("banana")
        # "ana" appears at positions 1 and 3
        self.assertEqual(substr, "ana")
        self.assertIn(1, positions)
        self.assertIn(3, positions)
        
        substr, positions = longest_repeated_substring("abcabcabc")
        self.assertEqual(substr, "abcabc")
    
    def test_longest_common_prefix(self):
        """Test longest common prefix."""
        self.assertEqual(longest_common_prefix("abcdef", "abcxyz"), 3)
        self.assertEqual(longest_common_prefix("xyz", "abc"), 0)
        self.assertEqual(longest_common_prefix("same", "same"), 4)
    
    def test_find_all_repeated_substrings(self):
        """Test finding all repeated substrings."""
        result = find_all_repeated_substrings("banana", min_length=2)
        self.assertTrue(len(result) > 0)
        # Check that substrings are sorted by length
        lengths = [len(s) for s, _ in result]
        self.assertEqual(lengths, sorted(lengths, reverse=True))


class TestPeriodDetection(unittest.TestCase):
    """Test period detection functions."""
    
    def test_find_minimal_period(self):
        """Test finding minimal period."""
        period = find_minimal_period("abcabcabc")
        self.assertEqual(period.period, 3)
        self.assertTrue(period.is_periodic)
        self.assertEqual(period.period_string, "abc")
        
        period = find_minimal_period("aaaa")
        self.assertEqual(period.period, 1)
        self.assertTrue(period.is_periodic)
        
        period = find_minimal_period("abcde")
        self.assertEqual(period.period, 5)
        self.assertFalse(period.is_periodic)
    
    def test_is_rotation(self):
        """Test rotation detection."""
        self.assertTrue(is_rotation("abcde", "cdeab"))
        self.assertTrue(is_rotation("abc", "abc"))
        self.assertFalse(is_rotation("abcde", "abced"))
        self.assertFalse(is_rotation("abc", "abcd"))
    
    def test_find_all_rotations(self):
        """Test finding all rotations."""
        rotations = find_all_rotations("abc")
        self.assertEqual(len(rotations), 3)
        self.assertIn("abc", rotations)
        self.assertIn("bca", rotations)
        self.assertIn("cab", rotations)
        
        # Periodic string should have fewer unique rotations
        rotations = find_all_rotations("aaaa")
        self.assertEqual(len(rotations), 1)


class TestPalindromeOperations(unittest.TestCase):
    """Test palindrome-related functions."""
    
    def test_longest_palindromic_prefix(self):
        """Test longest palindromic prefix."""
        self.assertEqual(longest_palindromic_prefix("abacaba"), "abacaba")
        self.assertEqual(longest_palindromic_prefix("abaxyz"), "aba")
        # Single character is a palindrome (length 1)
        self.assertEqual(longest_palindromic_prefix("xyz"), "x")
        self.assertEqual(longest_palindromic_prefix(""), "")
    
    def test_longest_palindromic_suffix(self):
        """Test longest palindromic suffix."""
        self.assertEqual(longest_palindromic_suffix("xyzaba"), "aba")
        self.assertEqual(longest_palindromic_suffix("abacaba"), "abacaba")
        # Single character is a palindrome (length 1)
        self.assertEqual(longest_palindromic_suffix("xyz"), "z")


class TestSimilarity(unittest.TestCase):
    """Test similarity functions."""
    
    def test_similarity_score(self):
        """Test similarity score calculation."""
        self.assertEqual(similarity_score("abcdef", "abcxyz"), 0.5)
        self.assertEqual(similarity_score("same", "same"), 1.0)
        self.assertEqual(similarity_score("abc", "xyz"), 0.0)
        self.assertEqual(similarity_score("", ""), 1.0)
        self.assertEqual(similarity_score("abc", ""), 0.0)
    
    def test_batch_similarity(self):
        """Test batch similarity calculation."""
        base = "abcdef"
        strings = ["abcxyz", "abcdef", "xyz", ""]
        scores = batch_similarity(base, strings)
        self.assertEqual(len(scores), 4)
        # "abcxyz" shares 3 chars with "abcdef"
        self.assertEqual(scores[1], 1.0)  # Exact match


class TestCompression(unittest.TestCase):
    """Test compression and decompression."""
    
    def test_compress_string(self):
        """Test string compression."""
        pattern, count = compress_string("abcabcabc")
        self.assertEqual(pattern, "abc")
        self.assertEqual(count, 3)
        
        pattern, count = compress_string("aaaa")
        self.assertEqual(pattern, "a")
        self.assertEqual(count, 4)
        
        pattern, count = compress_string("abcde")
        self.assertEqual(pattern, "abcde")
        self.assertEqual(count, 1)
    
    def test_decompress_string(self):
        """Test string decompression."""
        self.assertEqual(decompress_string("abc", 3), "abcabcabc")
        self.assertEqual(decompress_string("a", 4), "aaaa")
    
    def test_compress_decompress_roundtrip(self):
        """Test roundtrip compression."""
        original = "abcabcabc"
        pattern, count = compress_string(original)
        restored = decompress_string(pattern, count)
        self.assertEqual(restored, original)


class TestDistinctSubstringCount(unittest.TestCase):
    """Test distinct substring counting."""
    
    def test_distinct_substring_count(self):
        """Test counting distinct substrings."""
        # "abc" has 6 distinct substrings: a, b, c, ab, bc, abc
        self.assertEqual(distinct_substring_count("abc"), 6)
        
        # "aaaa" has 4 distinct substrings: a, aa, aaa, aaaa
        self.assertEqual(distinct_substring_count("aaaa"), 4)
        
        # Empty string has 0
        self.assertEqual(distinct_substring_count(""), 0)


class TestBorderConversion(unittest.TestCase):
    """Test Z-array to border array conversion."""
    
    def test_z_to_border(self):
        """Test Z to border conversion."""
        z = z_array("aabcaabxaaz")
        border = z_to_border(z)
        self.assertEqual(len(border), len(z))
    
    def test_border_to_z(self):
        """Test border to Z conversion."""
        z = z_array("abcabcabc")
        border = z_to_border(z)
        z_back = border_to_z(border)
        # Conversion might not be exact due to information loss
        self.assertEqual(len(z_back), len(z))


class TestZPatternMatcher(unittest.TestCase):
    """Test pattern matcher class."""
    
    def test_single_pattern(self):
        """Test matching single pattern."""
        matcher = ZPatternMatcher(["abc"])
        results = matcher.search("abcabc")
        self.assertEqual(len(results), 2)
        self.assertEqual(results[0], (0, 0, "abc"))
        self.assertEqual(results[1], (0, 3, "abc"))
    
    def test_multiple_patterns(self):
        """Test matching multiple patterns."""
        matcher = ZPatternMatcher(["ab", "bc"])
        results = matcher.search("abcabc")
        self.assertEqual(len(results), 4)  # 2 "ab" + 2 "bc"
    
    def test_search_first(self):
        """Test finding first occurrence."""
        matcher = ZPatternMatcher(["bc", "ab"])
        result = matcher.search_first("abcabc")
        self.assertIsNotNone(result)
        self.assertEqual(result[1], 0)  # "ab" appears first
    
    def test_count_all(self):
        """Test counting all patterns."""
        matcher = ZPatternMatcher(["a", "b", "c"])
        counts = matcher.count_all("abcabc")
        self.assertEqual(counts["a"], 2)
        self.assertEqual(counts["b"], 2)
        self.assertEqual(counts["c"], 2)
    
    def test_no_match(self):
        """Test when no pattern matches."""
        matcher = ZPatternMatcher(["xyz"])
        self.assertEqual(matcher.search("abcabc"), [])
        self.assertIsNone(matcher.search_first("abcabc"))


class TestUtilityFunctions(unittest.TestCase):
    """Test utility functions."""
    
    def test_visualize_z_array(self):
        """Test Z-array visualization."""
        visualization = visualize_z_array("aaaa")
        self.assertIn("String:", visualization)
        self.assertIn("aaaa", visualization)
        self.assertIn("Z:", visualization)
    
    def test_validate_z_array(self):
        """Test Z-array validation."""
        s = "aabcaabxaaz"
        z = z_array(s)
        self.assertTrue(validate_z_array(s, z))
        
        # Invalid Z-array
        invalid_z = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
        self.assertFalse(validate_z_array(s, invalid_z))
        
        # Wrong length
        self.assertFalse(validate_z_array(s, z[:5]))


class TestDataClasses(unittest.TestCase):
    """Test data classes."""
    
    def test_z_match(self):
        """Test ZMatch dataclass."""
        match = ZMatch(index=2, length=3, text="abcdef")
        self.assertEqual(match.index, 2)
        self.assertEqual(match.length, 3)
        self.assertEqual(match.matched_substring, "cde")
    
    def test_string_period(self):
        """Test StringPeriod dataclass."""
        period = StringPeriod(string="abcabc", period=3, is_periodic=True)
        self.assertEqual(period.period_string, "abc")


class TestEdgeCases(unittest.TestCase):
    """Test edge cases."""
    
    def test_single_char_string(self):
        """Test operations on single character."""
        self.assertEqual(z_array("a"), [0])
        self.assertEqual(find_all_occurrences("a", "a"), [0])
        self.assertEqual(longest_prefix_suffix("a"), 0)
    
    def test_all_unique_chars(self):
        """Test string with all unique characters."""
        s = "abcdefg"
        self.assertEqual(z_array(s), [0] * 7)
        self.assertEqual(longest_prefix_suffix(s), 0)
        period = find_minimal_period(s)
        self.assertFalse(period.is_periodic)
    
    def test_repeated_single_char(self):
        """Test string of repeated single character."""
        s = "aaaaaa"
        z = z_array(s)
        # Z[1] should be 5, Z[2] should be 4, etc.
        self.assertEqual(z[1], 5)
        self.assertEqual(z[2], 4)
        period = find_minimal_period(s)
        self.assertEqual(period.period, 1)
    
    def test_unicode_strings(self):
        """Test with Unicode strings."""
        s = "你好你好"
        self.assertEqual(find_all_occurrences("你好", s), [0, 2])
        period = find_minimal_period(s)
        self.assertEqual(period.period, 2)


if __name__ == '__main__':
    unittest.main(verbosity=2)